src/swap.ts

```diff
@@ -5,6 +5,16 @@ import { Field } from "./cnd/siren";
 import { sleep, timeoutPromise, TryParams } from "./util/timeout_promise";
 import { AllWallets, Wallets } from "./wallet";
 
+export class WalletError extends Error {
+  constructor(
+    public readonly attemptedAction: string,
+    public readonly source: Error,
+    public readonly callParams: any
+  ) {
+    super(JSON.stringify(source));
```

Comment:
:mega: Sorry for sending you on a wrong path here with the source. If JS Errors don't support being chained, we should just pass our own message and be done but definitely not JSON.stringifing the error type. That can lead to runtime errors if the type has self-references.